project-koku/koku
# [COST-3582] Use today if end_date is earlier than summary_month

## Jira Ticket

[COST-3582](https://issues.redhat.com/browse/COST-3582)

## Description

This change will add a check to make sure the end_date used is not earlier than the summary_month when getting list of month tuples in the `get_months_in_date_range` method. 
This pr also refactors the method to leverage @samdoran's suggestions [here](https://gist.github.com/samdoran/94bf7071186b58d4a888a68f32fc564f#file-utils-py-L26), to minimize datetime conversions and use more consistent datetime strings.

## Testing

1. Before checking out this branch, see that error is triggered by with an early report end date, by running the `get_months_in_date_range` in your django shell.

        ```        
        In [1]: from api.utils import get_months_in_date_range

        In [2]: report = {
        ...:     'end': '2022-05-22',
        ...:     'start': '2022-05-21',
        ...:     'invoice_month': None,
        ...:     'manifest_id': 272022,
        ...:     'provider_type': 'OCI',
        ...:     'provider_uuid': 'de3408b7-4781-4df2-9404-e16eba10aedd',
        ...:     'schema_name': 'acct610001,
        ...:     'tracing_id': 'de3108b7-4781-4df2-9404-e16eba10aedd:202302'
        ...: }

        In [3]: months = get_months_in_date_range(report)
        [2023-03-24 11:53:21,655] INFO None using start: 2022-05-21 and end: 2022-05-22 dates from manifest
        ---------------------------------------------------------------------------
        IndexError                                Traceback (most recent call last)

        ```


2. Checkout this Branch
3. Run the `get_months_in_date_range` in a new django shell and you should not get the `IndexError`
    ```
    In [18]: from api.utils import get_months_in_date_range

    In [19]: report = {
        ...:     'end': '2022-05-22',
        ...:     'start': '2022-05-21',
        ...:     'invoice_month': None,
        ...:     'manifest_id': 272022,
        ...:     'provider_type': 'OCI',
        ...:     'provider_uuid': 'de3408b7-4781-4df2-9404-e16eba10aedd',
        ...:     'schema_name': 'acct10001',
        ...:     'tracing_id': 'de3108b7-4781-4df2-9404-e16eba10aedd:202302'
        ...: }

    In [20]: months = get_months_in_date_range(report)
    [2023-03-24 11:52:40,892] INFO None using start: 2022-05-21 and end: 2022-05-22 dates from manifest
    ```

## Notes

...
